### ISSUE TITLE
Modifying mass with MERCURIUS

### ISSUE THREAD

**Issue by cl3425** [hannorein/rebound#449]
Hi! I would like to add mass loss to my many-particle simulations through REBOUNDX, but the MERCURIUS integrator does not support this or updater operators in general, at least from what I gathered from looking through the REBOUND and REBOUNDX code, as well as Tamayo et al 2019 and related papers.

What are the major hurdles here? Does it lie in the fact that calculations are split up (forcing a decision as to where to update positions and velocities); is it in the coordinate transformations to democratic-heliocentric (which mass changes to particles might complicate); or is it just that no one's gotten to it yet (in which case I'd be happy to help!)? Thanks for your time!

**Comment by hannorein** [comment]
It's complicated! Almost impossible to do it right, but potentially doable in one specific scenario. 

In the Mercurius integrator, the switching radius depends on the hill radius. If you change the mass of a particle, you need to update the hill radius. But the forces in Mercurius depend on the hill radius. To do this perfectly, you'd need to calculate the derivative of the masses and recalculate the equations of motions. Another issue is that the timestep in the adaptive part of the integrator. It might be tricky to synchronize all particles. Imagine a scenario where a particle's mass changes in one of the substeps. That could trigger changes in other particles which are not having a close encounter. 

All in all, I don't see a way forward to implement this "once and for all". if your mass change is very slow and adiabaitic, it might be not a problem and you can do something simple. Just check that the results are numerically converged. (FYI: I'm working on a different approach to these close encounter integrators. But it's not ready yet and I don't want to promise a time-line just yet)

**Comment by cl3425** [comment]
Thanks for your quick response, @hannorein! I'm starting to grasp how massive of an effort this code is. 

While the mass loss I'm simulating is not slow, I do have a case that I'm wondering might be set up to help me sidestep at least the first issue you listed: a single particle experiencing mass loss (ie. a star), with a handful of constant-mass planets and thousands of test particles. Given that case, what if the switching radius didn't depend on the Hill radius and instead followed one of the other three dcrit criteria given in reb_integrator_mercurius_part1(), like physical radius or some factor of it? The cost is greater false close encounters depending on how conservative I draw my radii, but maybe with only one mass-changing particle, these losses are minimized? 

To your second point, do the synchronization issues lie specifically in the un-combining of half-steps, or in the coordinate transformation, or both? If it's just in the former, is there a super_safe_mode where full Kepler steps are not allowed (as opposed to making the full step and then dialing it back...that might already be the case with safe_mode for Mercurius and, regardless, may be a moot point; I'm just having trouble reconciling the Mercurius code with the advanced WHFast tutorial's comments on synchronization)? 

Finally, I'm excited to see this new approach (no pressure :D)!

**Comment by hannorein** [comment]
How quickly does your star loose mass? Is it a few orbital periods (supernova style) or is it adiabatic (thousands of orbital periods). 

If you're in the super nova regime, I'd recommend just using IAS15 for that short phase. It might be a bit slow, but if you're only integrating for a few orbits, you can set the various control parameters (min_dt) such that IAS15 should be fast enough.

If you're in the adiabatic phase, you should not have an issue anyway. You can slowly change the mass, and then recalculate the coordinates, hill radii, etc, every now and then. I'd suggest running some sort of convergence study to make sure there's not numerical issue. 

Also, don't worry too much about the hill radius criteria. It will not be a formally symplectic integrator in the end, but you should be fine. You can also just make the initial switching radii large enough so that you'll be fine for the entire duration of the simulation. 

In short: I'd just set it up one way or the other and run several tests to make sure it works.

**Comment by cl3425** [comment]
My star is a sun-like star in the post-main sequence regime, so more like over ~1e5 or 1e6 orbits and definitely not supernova style. 

Good point! I'll try a battery of slow mass change tests and see how they converge. And so if I go with a non-Hill-radius-dependent switching radius, will that preserve symplecticity? I guess the other thing I could do is output energy errors to get a sense of how much that matters anyway. Thanks for your help, Hanno!

**Comment by hannorein** [comment]
It's not going to be a formally symplectic integrator. But that really doesn't matter in your case! When you monitor the energy error, keep in mind that the energy is not conserved whenever a mass changes. It might be most convincing if you do the convergence study on whatever physical quantity you're interested in. For example, you might want to show that the number of ejected particles is independent of purely numerical parameters.

**Comment by cl3425** [comment]
Ah, fair enough, I was fixated on the wrong thing! Thank you again for your help; I'll close the issue now.
